packages/toolkit/src/tests/createAsyncThunk.test-d.ts

```diff
@@ -0,0 +1,903 @@
+import type {
+  AsyncThunk,
+  SerializedError,
+  ThunkDispatch,
+  UnknownAction,
+} from '@reduxjs/toolkit'
+import {
+  configureStore,
+  createAsyncThunk,
+  createReducer,
+  createSlice,
+  unwrapResult,
+} from '@reduxjs/toolkit'
+
+import type { TSVersion } from '@phryneas/ts-version'
+import type { AxiosError } from 'axios'
+import apiRequest from 'axios'
+
+const defaultDispatch = (() => {}) as ThunkDispatch<{}, any, UnknownAction>
+const unknownAction = { type: 'foo' } as UnknownAction
+
+describe('type tests', () => {
+  test('basic usage', () => {
+    ;(async function () {
+      const asyncThunk = createAsyncThunk('test', (id: number) =>
+        Promise.resolve(id * 2),
+      )
+
+      const reducer = createReducer({}, (builder) =>
+        builder
+          .addCase(asyncThunk.pending, (_, action) => {
+            expectTypeOf(action).toEqualTypeOf<
+              ReturnType<(typeof asyncThunk)['pending']>
+            >()
+          })
+
+          .addCase(asyncThunk.fulfilled, (_, action) => {
+            expectTypeOf(action).toEqualTypeOf<
+              ReturnType<(typeof asyncThunk)['fulfilled']>
+            >()
+
+            expectTypeOf(action.payload).toBeNumber()
+          })
+
+          .addCase(asyncThunk.rejected, (_, action) => {
+            expectTypeOf(action).toEqualTypeOf<
+              ReturnType<(typeof asyncThunk)['rejected']>
+            >()
+
+            expectTypeOf(action.error).toMatchTypeOf<
+              Partial<Error> | undefined
+            >()
+          }),
+      )
+
+      const promise = defaultDispatch(asyncThunk(3))
+
+      expectTypeOf(promise.requestId).toBeString()
+
+      expectTypeOf(promise.arg).toBeNumber()
+
+      expectTypeOf(promise.abort).toEqualTypeOf<(reason?: string) => void>()
+
+      const result = await promise
+
+      if (asyncThunk.fulfilled.match(result)) {
+        expectTypeOf(result).toEqualTypeOf<
+          ReturnType<(typeof asyncThunk)['fulfilled']>
+        >()
+      } else {
+        expectTypeOf(result).toEqualTypeOf<
+          ReturnType<(typeof asyncThunk)['rejected']>
+        >()
+      }
+
+      promise
+        .then(unwrapResult)
+        .then((result) => {
+          expectTypeOf(result).toBeNumber()
+
+          expectTypeOf(result).not.toMatchTypeOf<Error>()
+        })
+        .catch((error) => {
+          // catch is always any-typed, nothing we can do here
+          expectTypeOf(error).toBeAny()
+        })
+    })()
+  })
+
+  test('More complex usage of thunk args', () => {
+    ;(async function () {
```

Comment:
this can be un-IIFE'd now it's got its own test block - same for the others in this file